### PR TITLE
YQ-2735 Hopping on analytic (v2)

### PR DESCRIPTION
**Pull request by kardymonds**
### Changelog entry 
Fixed reading from S3 with GROUP BY HOP() [kqp].

### Changelog category 
* Bugfix

### CODE REVIEW

File: ydb/library/yql/providers/s3/provider/yql_s3_datasource.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include <ydb/library/yql/providers/s3/proto/credentials.pb.h>
 
 #include <ydb/library/yql/utils/log/log.h>
+#include <ydb/library/yql/providers/dq/common/yql_dq_common.h>
```
Comment (UgnineSirdis) [inline ydb/library/yql/providers/s3/provider/yql_s3_datasource.cpp:15]: Why is this change?

File: ydb/library/yql/providers/s3/provider/yql_s3_datasink.cpp [hunk 2]
```diff
@@ -4,10 +4,14 @@
 #include <ydb/library/yql/core/expr_nodes/yql_expr_nodes.h>
 #include <ydb/library/yql/providers/s3/expr_nodes/yql_s3_expr_nodes.h>
 
+//#include <ydb/library/yql/providers/dq/provider/yql_dq_recapture.h>
+
 #include <ydb/library/yql/providers/common/provider/yql_provider.h>
 #include <ydb/library/yql/providers/common/provider/yql_provider_names.h>
 #include <ydb/library/yql/providers/common/provider/yql_data_provider_impl.h>
 
+#include <ydb/library/yql/providers/dq/common/yql_dq_common.h>
```
Comment (UgnineSirdis) [inline ydb/library/yql/providers/s3/provider/yql_s3_datasink.cpp:13]: Why is this change?

File: ydb/library/yql/providers/s3/provider/yql_s3_datasink.cpp [hunk 3]
```diff
@@ -4,10 +4,14 @@
 #include <ydb/library/yql/core/expr_nodes/yql_expr_nodes.h>
 #include <ydb/library/yql/providers/s3/expr_nodes/yql_s3_expr_nodes.h>
 
+//#include <ydb/library/yql/providers/dq/provider/yql_dq_recapture.h>
```
Comment (UgnineSirdis) [inline ydb/library/yql/providers/s3/provider/yql_s3_datasink.cpp:7]: If this header is not needed, it's better to remove it, not comment.

File: ydb/library/yql/providers/s3/provider/ya.make [hunk 4]
```diff
@@ -43,6 +43,7 @@ PEERDIR(
     ydb/library/yql/providers/common/token_accessor/client
     ydb/library/yql/providers/common/transform
     ydb/library/yql/providers/dq/common
+    ydb/library/yql/providers/dq/provider
```
Comment (UgnineSirdis) [inline ydb/library/yql/providers/s3/provider/ya.make:46]: Why is this change? This seems to be not proper dependency

File: ydb/library/yql/providers/dq/common/yql_dq_common.h [hunk 5]
```diff
@@ -4,6 +4,7 @@
 
 #include <util/generic/string.h>
 #include <map>
+#include <ydb/library/yql/providers/s3/provider/yql_s3_provider.h>
```
Comment (UgnineSirdis) [inline ydb/library/yql/providers/dq/common/yql_dq_common.h:7]: dq provider must not depend on s3 provider

**Review by UgnineSirdis — CHANGES_REQUESTED** [review]

File: ydb/library/yql/dq/opt/dq_opt_hopping.cpp [hunk 6]
```diff
@@ -0,0 +1,794 @@
...
+            .Output()
+                .Stage(stage)
+                .Index().Build(0)
+                .Build()
+            .Done();
+    }
+}
+
+
+} // NYql::NDqs
```
Comment (uzhastik) [inline ydb/library/yql/dq/opt/dq_opt_hopping.cpp:794]: wrong namespace name

File: ydb/library/yql/providers/dq/common/yql_dq_common.cpp [hunk 7]
```diff
@@ -13,6 +13,15 @@
 
 #include <util/string/split.h>
 
+#include <ydb/library/yql/dq/opt/dq_opt_log.h>
+#include "yql_dq_settings.h"
```
Comment (UgnineSirdis) [inline ydb/library/yql/providers/dq/common/yql_dq_common.cpp:17]: According to arcadia style this header should be at the beginning

File: ydb/library/yql/providers/dq/common/yql_dq_common.cpp [hunk 8]
```diff
@@ -135,4 +144,14 @@ bool NeedFallback(const NDq::TEvDq::TEvAbortExecution::TPtr& ev) {
...
 
 } // namespace NCommon
+
+using namespace NNodes;
+namespace NDq {
+    bool CheckJoinColumns(const TExprBase& node);
+} // namespace NDq
+
+
+const THashSet<TStringBuf> VALID_SOURCES = {DqProviderName, ConfigProviderName, YtProviderName, ClickHouseProviderName, YdbProviderName, S3ProviderName, PgProviderName};
```
Comment (UgnineSirdis) [inline ydb/library/yql/providers/dq/common/yql_dq_common.cpp:154]: Unused static variables

File: ydb/library/yql/providers/dq/common/yql_dq_common.cpp [hunk 9]
```diff
@@ -135,4 +144,14 @@ bool NeedFallback(const NDq::TEvDq::TEvAbortExecution::TPtr& ev) {
 }
 
 } // namespace NCommon
+
+using namespace NNodes;
```
Comment (UgnineSirdis) [inline ydb/library/yql/providers/dq/common/yql_dq_common.cpp:148]: Do we really need this at the end of the file?

File: ydb/library/yql/providers/dq/common/yql_dq_common.cpp [hunk 10]
```diff
@@ -135,4 +144,14 @@ bool NeedFallback(const NDq::TEvDq::TEvAbortExecution::TPtr& ev) {
...
+namespace NDq {
+    bool CheckJoinColumns(const TExprBase& node);
```
Comment (UgnineSirdis) [inline ydb/library/yql/providers/dq/common/yql_dq_common.cpp:150]: Where is this function implemented? I can't see

File: ydb/library/yql/providers/dq/common/yql_dq_common.h [hunk 11]
```diff
@@ -19,4 +19,5 @@ bool IsRetriable(const NDq::TEvDq::TEvAbortExecution::TPtr& ev);
 bool NeedFallback(NYql::NDqProto::StatusIds::StatusCode statusCode);
 bool NeedFallback(const NDq::TEvDq::TEvAbortExecution::TPtr& ev);
 } // namespace NCommon
+
```
Comment (UgnineSirdis) [inline ydb/library/yql/providers/dq/common/yql_dq_common.h:22]: Why do we need empty string here?

File: ydb/library/yql/dq/opt/dq_opt_hopping.h [hunk 12]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include <ydb/library/yql/dq/expr_nodes/dq_expr_nodes.h>
+
+#include <util/datetime/base.h>
+#include <util/generic/ptr.h>
```
Comment (UgnineSirdis) [inline ydb/library/yql/dq/opt/dq_opt_hopping.h:6]: I can't see the direct usage of this header here

File: ydb/core/kqp/opt/logical/kqp_opt_log.cpp [hunk 13]
```diff
@@ -109,8 +110,28 @@ class TKqpLogicalOptTransformer : public TOptimizeTransformerBase {
     }
 
     TMaybeNode<TExprBase> RewriteAggregate(TExprBase node, TExprContext& ctx) {
-        TExprBase output = DqRewriteAggregate(node, ctx, TypesCtx, false, KqpCtx.Config->HasOptEnableOlapPushdown() || KqpCtx.Config->HasOptUseFinalizeByKey(), KqpCtx.Config->HasOptUseFinalizeByKey());
-        DumpAppliedRule("RewriteAggregate", node.Ptr(), output.Ptr(), ctx);
+        TMaybeNode<TExprBase> output;
+        auto aggregate = node.Cast<TCoAggregateBase>();
+        auto input = aggregate.Input().Maybe<TDqConnection>();
```
Comment (UgnineSirdis) [inline ydb/core/kqp/opt/logical/kqp_opt_log.cpp:115]: We can place this string after `if (hopSetting) {`, because it is not used outside it

**Review by UgnineSirdis — CHANGES_REQUESTED** [review]

**Review by UgnineSirdis — DISMISSED** [review]

File: ydb/library/yql/dq/opt/dq_opt_hopping.cpp [hunk 14]
```diff
@@ -0,0 +1,794 @@
...
+    TExprContext& ctx,
+    bool analyticsMode,
+    bool defaultWatermarksMode,
+    bool asyncActor)
+{
+    const bool enableWatermarks = !analyticsMode &&
+        defaultWatermarksMode &&
+        hoppingTraits.Version().Cast<TCoAtom>().StringValue() == "v2";
+    if (enableWatermarks && asyncActor) {
+        ctx.AddError(TIssue(ctx.GetPosition(aggregate.Pos()), "Watermarks should be used only with async compute actor"));
```
Comment (resetius) [inline ydb/library/yql/dq/opt/dq_opt_hopping.cpp:647]: Maybe 'Watermarks should be used only with sync compute actor' ?

File: ydb/library/yql/providers/dq/opt/logical_optimize.cpp [hunk 15]
```diff
@@ -336,7 +310,13 @@ class TDqsLogicalOptProposalTransformer : public TOptimizeTransformerBase {
         auto hopSetting = GetSetting(aggregate.Settings().Ref(), "hopping");
         if (input) {
             if (hopSetting) {
-                return RewriteAsHoppingWindow(node, ctx, input.Cast());
+                bool analyticsHopping = Config->AnalyticsHopping.Get().GetOrElse(false);
+                const auto lateArrivalDelay = TDuration::MilliSeconds(Config->WatermarksLateArrivalDelayMs
+                    .Get()
+                    .GetOrElse(TDqSettings::TDefault::WatermarksLateArrivalDelayMs));
+                bool defaultWatermarksMode = Config->WatermarksMode.Get() == "default";
+                bool asyncActor = Config->ComputeActorType.Get() != "async";
```
Comment (resetius) [inline ydb/library/yql/providers/dq/opt/logical_optimize.cpp:318]: Don't understand that. Maybe asyncActor = Config->ComputeActorType.Get() == "async"; ?

**Comment by resetius** [comment]
Pls don't do refactoring and fix/new_feature in one PR next time!

File: ydb/library/yql/providers/dq/opt/logical_optimize.cpp (hunk 15 above)
Comment (kardymonds) [reply ydb/library/yql/providers/dq/opt/logical_optimize.cpp]: Fixed

File: ydb/library/yql/dq/opt/dq_opt_hopping.cpp (hunk 14 above)
Comment (kardymonds) [reply ydb/library/yql/dq/opt/dq_opt_hopping.cpp:647]: Fixed

**Comment by kardymonds** [comment]
@resetius Fixed async/sync, look please again

**Review by resetius — DISMISSED** [review]

**Review by uzhastik — DISMISSED** [review]

**Review by UgnineSirdis — DISMISSED** [review]

**Review by UgnineSirdis — APPROVED** [review]

**Review by resetius — APPROVED** [review]

**Pull request merged**